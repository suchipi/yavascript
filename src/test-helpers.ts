import { spawn, Options as SpawnOptions } from "first-base";
import child_process from "child_process";
import path from "path";
import print from "@suchipi/print";
import * as inspectOptions from "./inspect-options";

export const TMP = child_process
  .execSync("realpath /tmp", { encoding: "utf-8" })
  .trim();

export const rootDir = (...parts: Array<string>) =>
  path.resolve(__dirname, "..", ...parts);

function getBinaryPath(platform: string, arch: string) {
  if (platform === "win32") {
    return path.resolve(__dirname, "..", "bin", "windows", "yavascript.exe");
  } else if (platform === "darwin") {
    if (arch.startsWith("arm")) {
      return path.resolve(__dirname, "..", "bin", "darwin-arm", "yavascript");
    } else {
      return path.resolve(__dirname, "..", "bin", "darwin", "yavascript");
    }
  } else if (platform === "linux") {
    return path.resolve(__dirname, "..", "bin", "linux", "yavascript");
  } else {
    throw new Error("Unsupported platform: " + platform);
  }
}

var binaryPath = getBinaryPath(
  process.platform,
  process.env.YS_TESTS_ARCH || process.arch
);

export { binaryPath, getBinaryPath };

export type EvaluateResult = {
  stdout: string;
  stderr: string;
  code: number | null;
  error: boolean;
};

export async function evaluate(
  code: string,
  options: SpawnOptions = {}
): Promise<EvaluateResult> {
  const runContext = spawn(binaryPath, ["-e", code], options);
  await runContext.completion;
  return runContext.result;
}

export function inspect(value: any): string {
  // options to inspect here match what is given to console.log
  return print(value, inspectOptions.forPrint);
}

export function cleanOutput(input: string): string {
  return input
    .replace(/yavascript:\d+/g, "yavascript")
    .replace(/\/bin\/[^/]+\/yavascript/g, "/bin/.../yavascript")
    .replace(new RegExp(rootDir(), "g"), "<rootDir>");
}

export function cleanResult(input: EvaluateResult): EvaluateResult {
  return {
    ...input,
    stdout: cleanOutput(input.stdout),
    stderr: cleanOutput(input.stderr),
  };
}
