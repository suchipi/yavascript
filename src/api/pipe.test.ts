import { evaluate } from "../test-helpers";

test("string -> new Uint8Array", async () => {
  const input = { data: "Hello, world!" };
  const result = await evaluate(`pipe(${JSON.stringify(input)}, Uint8Array)`);
  expect(result).toMatchInlineSnapshot(`
    {
      "code": 0,
      "error": false,
      "stderr": "",
      "stdout": "{
      bytesTransferred: 13
      target: Uint8Array [
        │0x00000000│ 48 65 6C 6C 6F 2C 20 77 6F 72 6C 64 21
      ]
    }
    ",
    }
  `);
});

test("string -> existing Uint8Array (exact size)", async () => {
  const input = { data: "Hello, world!" };
  const outputBytes = [
    0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x2c, 0x20, 0x77, 0x6f, 0x72, 0x6c, 0x64,
    0x21,
  ];

  const result = await evaluate(
    `pipe(${JSON.stringify(input)}, new Uint8Array(${outputBytes.length}))`
  );

  expect(result).toMatchInlineSnapshot(`
    {
      "code": 0,
      "error": false,
      "stderr": "",
      "stdout": "{
      bytesTransferred: 13
      target: Uint8Array [
        │0x00000000│ 48 65 6C 6C 6F 2C 20 77 6F 72 6C 64 21
      ]
    }
    ",
    }
  `);
});

test("string -> existing Uint8Array (smaller; truncates)", async () => {
  const input = { data: "Hello, world!" };
  const outputBytes = [
    0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x2c, 0x20, 0x77, 0x6f, 0x72, 0x6c, 0x64,
    0x21,
  ];

  const result = await evaluate(
    `pipe(${JSON.stringify(input)}, new Uint8Array(${outputBytes.length - 4}))`
  );

  expect(result).toMatchInlineSnapshot(`
    {
      "code": 0,
      "error": false,
      "stderr": "",
      "stdout": "{
      bytesTransferred: 9
      target: Uint8Array [
        │0x00000000│ 48 65 6C 6C 6F 2C 20 77 6F
      ]
    }
    ",
    }
  `);
});

test("string -> existing Uint8Array (larger; overwrites but leaves existing data alone)", async () => {
  const input = { data: "Hello, world!" };
  const outputBytes = [
    0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x2c, 0x20, 0x77, 0x6f, 0x72, 0x6c, 0x64,
    0x21,
  ];

  const result = await evaluate(
    `pipe(${JSON.stringify(input)}, new Uint8Array(new Array(${
      outputBytes.length + 4
    }).fill(1)))`
  );

  expect(result).toMatchInlineSnapshot(`
    {
      "code": 0,
      "error": false,
      "stderr": "",
      "stdout": "{
      bytesTransferred: 13
      target: Uint8Array [
        │0x00000000│ 48 65 6C 6C 6F 2C 20 77 6F 72 6C 64 21 01 01 01
        │0x00000010│ 01
      ]
    }
    ",
    }
  `);
});

// TODO: every other combination of input/output...

test("can handle NUL byte in content", async () => {
  // NES ROM Header :3
  const input = [
    0x4e, 0x45, 0x53, 0x1a, 0x02, 0x02, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00,
  ];

  const result = await evaluate(
    `pipe(new Uint8Array(${JSON.stringify(input)}), Uint8Array)`
  );

  expect(result).toMatchInlineSnapshot(`
    {
      "code": 0,
      "error": false,
      "stderr": "",
      "stdout": "{
      bytesTransferred: 16
      target: Uint8Array [
        │0x00000000│ 4E 45 53 1A 02 02 11 00 00 00 00 00 00 00 00 00
      ]
    }
    ",
    }
  `);
});
