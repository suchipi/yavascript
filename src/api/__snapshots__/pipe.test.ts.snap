// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can handle NUL byte in content 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  bytesTransferred: 16
  target: Uint8Array [
    │0x00000000│ 4E 45 53 1A 02 02 11 00 00 00 00 00 00 00 00 00
  ]
}
",
}
`;

exports[`string -> existing Uint8Array (exact size) 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  bytesTransferred: 13
  target: Uint8Array [
    │0x00000000│ 48 65 6C 6C 6F 2C 20 77 6F 72 6C 64 21
  ]
}
",
}
`;

exports[`string -> existing Uint8Array (larger; overwrites but leaves existing data alone) 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  bytesTransferred: 13
  target: Uint8Array [
    │0x00000000│ 48 65 6C 6C 6F 2C 20 77 6F 72 6C 64 21 01 01 01
    │0x00000010│ 01
  ]
}
",
}
`;

exports[`string -> existing Uint8Array (smaller; truncates) 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  bytesTransferred: 9
  target: Uint8Array [
    │0x00000000│ 48 65 6C 6C 6F 2C 20 77 6F
  ]
}
",
}
`;

exports[`string -> new Uint8Array 1`] = `
{
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  bytesTransferred: 13
  target: Uint8Array [
    │0x00000000│ 48 65 6C 6C 6F 2C 20 77 6F 72 6C 64 21
  ]
}
",
}
`;
